### PR TITLE
HPCC-31396 Introduce a class for tracking the lifetime of a span

### PR DESCRIPTION
**Pull request by ghalliday**
## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [x] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by ghalliday** [comment]
@rpastrana I think if the thread context changes work out there will be no need to store the ISpan in the logctx.  However that is a few changes away still.

**Comment by ghalliday** [comment]
Also, not currently tested.  Pushed to ensure it compiles and to run basic regression tests.  Will test properly tomorrow.

File: esp/platform/esp.hpp [hunk 1]
```diff
@@ -235,15 +235,13 @@ interface IEspContext : extends IInterface
     virtual void setRequest(IHttpMessage* req) = 0;
     virtual IHttpMessage* queryRequest() = 0;
 
-    virtual void setActiveSpan(ISpan * span)=0; // Only call this function directly if this object's lifetime matches the lifetime of the span.  If there is any doubt use EspContextSpanScope(ctx, span)
+    virtual void setServerSpan(ISpan * span)=0;     // Call this function to set the server span for the query.  The spans's lifetime will match the lifetime of the context object.
```
Comment (rpastrana) [inline esp/platform/esp.hpp:238]: I would consider renaming to setRequestSpan, it's a server span, but we want it to match the request's lifespan.

File: esp/platform/espcontext.cpp [hunk 2]
```diff
@@ -626,7 +626,7 @@ class CEspContext : public CInterface, implements IEspContext
     {
         return m_request;
     }
-    virtual void setActiveSpan(ISpan * span) override
+    virtual void setServerSpan(ISpan * span) override
     {
         m_activeSpan.set(span);
```
Comment (rpastrana) [inline esp/platform/espcontext.cpp:631]: do we still want to refer to this as active span or server or request span?

**Review by rpastrana — COMMENTED** [review]
@ghalliday looks good. left a couple of refactoring suggestions, but I didn't see any cppunit changes, new cases.

**Comment by ghalliday** [comment]
@rpastrana I renamed the function, and the member.  Also changed regression tests to use the new class.

**Review by rpastrana — APPROVED** [review]
@ghalliday looks good.
What's a good use case for explicitly calling endSpan?
I was hoping to add a test case illustrating/exercising that feature.

**Comment by ghalliday** [comment]
> @ghalliday looks good. What's a good use case for explicitly calling endSpan? I was hoping to add a test case illustrating/exercising that feature.

The main use case for calling endSpan is from the OwnedSpanScope - for the case where another object has kept a link to the span.  I'll see if I can add a test case.

**Pull request merged**